### PR TITLE
Feature/issue 241/duration time

### LINKED ISSUE #241: fast appointment end time, suggest the appointment duration time.
When a new appointment is being made and the start time is set up

We can choose scheduling for 15, 30 or 60 minutes, otherwise just set manually the end time of the meeting.

### PR DESCRIPTION
**Pull request by JessicaBuenfilSolis**
Closes #241

### CODE REVIEW

**Comment by JessicaBuenfilSolis** [comment]
#241

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 1]
```diff
@@ -122,11 +132,24 @@ class BookingFormComponent extends React.Component {
...
 
   setBookingStartTime = startTime => {
-    this.setState({ startTime }, () => this.enableEndTimeSelect());
+    if (startTime.hour && startTime.minute) {
+      this.setState({ timesDurationDisabled: false });
+    }
+    this.setState({ startTime }, () =>
+      this.setState(
+        prevState => this.updateMeetingDuration(prevState),
+        () => this.enableEndTimeSelect()
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: you'r just forwarding the call, you could express the same like this:

```js
this.setState(
        this.updateMeetingDuration,
        this.enableEndTimeSelect
)
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 2]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
     this.enableConferenceSelect();
   };
 
+  updateMeetingDuration = ({ startTime, endTime, timesMeetingDuration }) => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: move function out of component:

```jsx
// rename `timesMeetingDuration` to just `meetingDuration`
const getMeetingDuration = ({ startTime, endTime, meetingDuration }) => {
  // ... the rest of your code
  // just return times
  return times
}
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 3]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+        minute: startTime.minute
+      }),
+      "minute"
+    );
+
+    const times = [...timesMeetingDuration].map(element => {
+      if (element.value === minutes) {
+        element.selected = true;
+      } else {
+        element.selected = false;
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: you could express this like the following:

```jsx
const times = timesMeetingDuration.map(element => {
  return element.value === minutes ? { ...element, selected: true } : { ...elements, selected: false };
})
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 4]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+      }
+      return element;
+    });
+
+    return { timesMeetingDuration: times };
+  };
+
+  clearMeetingDuration = () => {
+    this.setState(({ timesMeetingDuration }) => {
+      const times = [...timesMeetingDuration].map(element => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: extract logic from `setState` into a separate function

```jsx
const clearSelections = times => {
  return times.map(element => ({ ...element, selected: false }));
}

// ... some code
clearMeetingDuration = () => {
  this.setState({ timesMeetingDuration }) => {
    const times = clearSelections(timesMeetingDuration);
    return { ...prevState, timesMeetingDuration: times };
  })
}
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 5]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+        element.selected = false;
+        return element;
+      });
+      return { timesMeetingDuration: times };
+    });
+  };
+
+  setTimeDurationSelected = time => {
+    const computedEndtime = moment({
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: extract logic into a separate function:

```js
const getEndTime = minutes = {
    const endTime = moment({
      hour: this.state.startTime.hour,
      minute: this.state.startTime.minute
    }).add(minutes, "minutes");

  return { hour: endTime.hour(), addZeros(endTime.minute())  };
}

// .. more code
setTimeDurationSelected = time => {
  const endTime = getEndTime(time.value);
  // rest of your code
}
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 6]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+    }).add(time.value, "minutes");
+
+    const endTime = {
+      hour: computedEndtime.hour(),
+      minute: addZeros(computedEndtime.minute())
+    };
+
+    this.setState({ endTime }, () =>
+      this.setState(({ timesMeetingDuration }) => {
+        const times = [...timesMeetingDuration].map(element => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: use `clearSelections`

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 7]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+          element.selected = false;
+          return element;
+        });
+        const idx = times.indexOf(time);
+        times[idx].selected = true;
+        return { timesMeetingDuration: times };
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]:
```jsx
return { ...prevState, timesMeetingDuration: times };
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 8]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+    return { timesMeetingDuration: times };
+  };
+
+  clearMeetingDuration = () => {
+    this.setState(({ timesMeetingDuration }) => {
+      const times = [...timesMeetingDuration].map(element => {
+        element.selected = false;
+        return element;
+      });
+      return { timesMeetingDuration: times };
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]:
```jsx
return { ...prevState, timesMeetingDuration: times };
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 9]
```diff
@@ -136,12 +159,72 @@ class BookingFormComponent extends React.Component {
...
+        element.selected = false;
+      }
+      return element;
+    });
+
+    return { timesMeetingDuration: times };
+  };
+
+  clearMeetingDuration = () => {
+    this.setState(({ timesMeetingDuration }) => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: favor `prevState` when using `this.setState()` over object destructuring

File: src/components/Modals/CreateMeeting/TimeSelect.js [hunk 10]
```diff
@@ -17,127 +17,76 @@ const styles = theme => ({
...
-        </Grid>
+  return (
+    <Grid container direction="row">
+      <Grid item xs={6}>
+        <FormControl className={classes.form}>
+          <InputLabel>Hour</InputLabel>
+          <Select
+            value={hour}
+            onChange={e => {
+              handleOnChangeHour(e, minute, onChangeHour);
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/TimeSelect.js]: change to 

```jsx
onChange={handleChangeHour(minute)}
```

File: src/components/Modals/CreateMeeting/TimeSelect.js [hunk 11]
```diff
@@ -17,127 +17,76 @@ const styles = theme => ({
...
+}
 
-  render() {
-    const { classes } = this.props;
-    let hours = this.state.hours.map(hour => (
-      <MenuItem value={hour[0]} key={cuid()}>
-        {hour[0]}
-      </MenuItem>
-    ));
+function handleOnChangeMinute(e, hour, onChangeMinute) {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/TimeSelect.js]: use currying:

```jsx
const handleChangeMinute = hour = event => {
  onChangeMinute({ hour, minute: event.target.value });
}
```

File: src/components/Modals/CreateMeeting/TimeSelect.js [hunk 12]
```diff
@@ -17,127 +17,76 @@ const styles = theme => ({
...
-      if (!this.state.hourSelected) {
-        this.setState({
-          hourSelected: this.props.endTime.hour,
-          minuteSelected: this.props.endTime.minute,
-          disabledMinutes: false
-        });
-      }
-    }
-  }
+function handleOnChangeHour(e, minute, onChangeHour) {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/TimeSelect.js]: change to:

```jsx
const handleChangeHour = minute = event => {
  onChangeHour({ hour: event.target.value, minute });
}
```

File: src/components/Modals/CreateMeeting/TimeSelect.js [hunk 13]
```diff
@@ -17,127 +17,76 @@ const styles = theme => ({
...
-    );
-  }
-}
+      <Grid item xs={6}>
+        <FormControl className={classes.form}>
+          <InputLabel>Minutes</InputLabel>
+          <Select
+            value={minute}
+            onChange={e => {
+              handleOnChangeMinute(e, hour, onChangeMinute);
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/TimeSelect.js]: use currying:

```jsx
onChange={handleChangeMinute(hour)}
```
```

File: src/components/Modals/CreateMeeting/MeetingDuration.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+const styles = theme => ({
+  timeSelector: {},
+  checked: {
+    color: "#0000FF"
+  }
+});
+
+const MeetingDuration = ({ classes, time, onSelectItem, disabled }) => {
+  const timeSelectedHandler = time => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/MeetingDuration.js]: user currying and rename:

```jsx
const handleClick = time => () => {
  onSelectItem(time);
}
```

File: src/components/Modals/CreateMeeting/MeetingDuration.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+        <Radio
+          disabled={disabled}
+          name="time duration selector"
+          value={time.value}
+          checked={time.selected}
+          color="default"
+          className={time.selected ? classes.checked : classes.timeSelector}
+          label={time.value}
+          onClick={() => {
+            timeSelectedHandler(time);
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/MeetingDuration.js]: use currying:
```jsx
onClick={handleClick(time)}
```

**Review by mishelashala — CHANGES_REQUESTED** [review]

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 16]
```diff
@@ -85,6 +85,46 @@ const styles = theme => ({
...
+    minute: endTime.minute
+  }).diff(
+    moment({
+      hour: startTime.hour,
+      minute: startTime.minute
+    }),
+    "minute"
+  );
+
+  const times = [...meetingDuration].map(element => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: unnecessary variable, just return the expression

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 17]
```diff
@@ -85,6 +85,46 @@ const styles = theme => ({
...
+
+  return times;
+};
+
+const clearSelections = times => {
+  return times.map(element => ({ ...element, selected: false }));
+};
+
+const getEndTime = (minutes, startTime) => {
+  const endTime = moment({
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: create another function `addMinutes`

```tsx
const addMinutes = (minutes, time) => {
  // do you need to create another moment object?
  // why not simply pass time?
  return moment({
    hour: time.hour,
    minute: time.minute
  }).add(minutes, "minutes");
}

// more code...
const endTime = addMinutes(minutes, endTime);
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 18]
```diff
@@ -132,23 +172,24 @@ class BookingFormComponent extends React.Component {
   };
 
   setBookingStartTime = startTime => {
+    console.log(startTime);
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: remove log

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 19]
```diff
@@ -132,23 +172,24 @@ class BookingFormComponent extends React.Component {
...
     if (startTime.hour && startTime.minute) {
       this.setState({ timesDurationDisabled: false });
     }
     this.setState({ startTime }, () =>
-      this.setState(
-        prevState => this.updateMeetingDuration(prevState),
-        () => this.enableEndTimeSelect()
-      )
+      this.setState(prevState => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: you don't need  a nested `setState`:

```tsx
this.setState(prevState => {
  const { startTime, endTime, meetingDuration } = prevState;
  const updatedMeetingDuration = getMeetingDuration({ meetingDuration, startTime, endTime });
  return { ...prevState, meetingDuration: updatedMeetingDuration };
})
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 20]
```diff
@@ -132,23 +172,24 @@ class BookingFormComponent extends React.Component {
...
     );
   };
 
   setBookingEndTime = endTime => {
     this.setState({ endTime }, () =>
-      this.setState(
-        prevState => this.updateMeetingDuration(prevState),
-        () => this.verifyQuickAppointment()
-      )
+      this.setState(prevState => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: same here, you don't need a nested `setState`

```tsx
this.setState(prevState => {
  const { startTime, endTime, meetingDuration } = prevState;
  const updatedMeetingDuration = getMeetingDuration({ meetingDuration, startTime, endTime });
  return { ...prevState, meetingDuration: updatedMeetingDuration };
})
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 21]
```diff
@@ -159,60 +200,25 @@ class BookingFormComponent extends React.Component {
...
-      hour: this.state.startTime.hour,
-      minute: this.state.startTime.minute
-    }).add(time.value, "minutes");
-
-    const endTime = {
-      hour: computedEndtime.hour(),
-      minute: addZeros(computedEndtime.minute())
-    };
-
+    const endTime = getEndTime(time.value, this.state.startTime);
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: this variable gets shadowed in the next 2 lines, do you need it here?

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 22]
```diff
@@ -75,9 +79,52 @@ const styles = theme => ({
...
+    textTransform: "lowercase"
   }
 });
 
+const getMeetingDuration = prevState => {
+  const { startTime, endTime, meetingDuration } = prevState;
+  if (!endTime.minute) {
+    return meetingDuration;
+  }
+  const minutes = moment({
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: move into a function called `getMinutesDiff`

```tsx
const getMinutesDiff = (startTime, endTime) => {
  // do you need to create another moment object?
  // why not simply pass endTime?
  return moment({
    hour: endTime.hour,
    minute: endTime.minute
  })
  .diff(
    // do you need to create another moment object?
    // why not simply pass startTime?
    moment({
      hour: startTime.hour,
      minute: startTime.minute
    }),
    "minute"
  );
}

const minutes = getMinutesDiff(startTime, endTime);
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 23]
```diff
@@ -136,12 +200,37 @@ class BookingFormComponent extends React.Component {
...
+  };
+
   setRoom = room => {
     this.setState({ roomId: room }, () => this.enableNextButton());
   };
 
   setDate = date => {
-    this.setState({ date: formatDashedDate(date) }, () => this.enableStartTimeSelect());
+    this.setState({ date: formatDashedDate(date) }, () =>
+      this.enableStartTimeSelect()
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: you don't need the extra function:

```tsx
this.setState({ date: formatDashedDate(date) }, this.enableStartTimeSelect)
```

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 24]
```diff
@@ -136,12 +200,37 @@ class BookingFormComponent extends React.Component {
...
+      const { meetingDuration } = prevState;
+      const times = clearSelections(meetingDuration);
+      return { ...prevState, meetingDuration: times };
+    });
+  };
+
+  setTimeDurationSelected = time => {
+    const endTime = getEndTime(time.value, this.state.startTime);
+    this.setState({ endTime }, () =>
+      this.setState(prevState => {
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: you don't need the nested `setState`

```tsx
this.setState(({ meetingDuration }) => {
  const times = clearSelections(meetingDuration);
  const idx = times.findIndex(element => {
    return element.value === time.value;
  });
  // add a validation for the index,
  // if no element it's found then the idx is
  // equal to -1
  times[idx].selected = true;
  return { ...prevState, meetingDuration: times };
})
```

**Review by mishelashala — CHANGES_REQUESTED** [review]

File: src/components/Modals/CreateMeeting/BookingForm.jsx [hunk 25]
```diff
@@ -121,12 +165,39 @@ class BookingFormComponent extends React.Component {
     this.setState({ disabledNextButton: false });
   };
 
-  setBookingStartTime = startTime => {
-    this.setState({ startTime }, () => this.enableEndTimeSelect());
+  setBookingStartTime = updateStartTime => {
+    if (updateStartTime.hour && updateStartTime.minute) {
+      this.setState({ timesDurationDisabled: false });
```
Comment (mishelashala) [inline src/components/Modals/CreateMeeting/BookingForm.jsx]: no conditionals `setState`

```tsx
this.setState(prevState => {
  const timesDurationDisabled = !(updateStartTime.hour && updateStartTime.minute);
  // ...
  return {
     ...prevState,
     startTime: updateStartTime,
     meetingDuration: updatedMeetingDuration,
     timesDurationDisabled
  };
})
```

**Review by mishelashala — CHANGES_REQUESTED** [review]

**Review by mishelashala — APPROVED** [review]

**Pull request merged**